### DISCUSSION TITLE
Consistency and transactions in DuckDB

### DISCUSSION THREAD

**Discussion by Kodiologist** [duckdb/duckdb#3371] [Q&A] [answered]
I'm a long-time user of SQLite and I'm interested in what cases DuckDB can effectively replace SQLite. In particular, I use SQLite for its hard guarantees of consistency and atomicity, so that (a) I can group my changes into transactions, so that if the client process dies mid-transaction, the database will be left just as it was before any of the changes, and (b) I can safely have several processes use the same database at once (although of course SQLite can't handle concurrent write attempts as fast as a traditional DMBS, since it has to lock the database for writing).

What kind of consistency guarantees does DuckDB provide? How do transactions work? For example, if your R code is inside a transaction and then your R program ends with an error, is the transaction rolled back? Forgive me if these things are explained in the documentation somewhere. All I could find was a passing mention of `BEGIN TRANSACTION … COMMIT` on [this page](https://duckdb.org/docs/data/insert) and the sentence "DuckDB provides transactional guarantees (ACID properties) through our custom, bulk-optimized Multi-Version Concurrency Control (MVCC)." on [this page](https://duckdb.org/why_duckdb).

**Comment by Alex-Monahan** [comment, marked as answer]
Hello and welcome! 

I can help with a few answers, but I am not a core developer so I don't know all of the details.

DuckDB does support transactional guarantees. We use a write ahead log (WAL) like SQLite frequently does. So, for question a, yes DuckDB guarantees transactional integrity. 

For question b, DuckDB has made a different design decision than SQLite. SQLite has to re read from disk for each query, which is slow for analytics. However, it allows them to have readers and a single writer at the same time. DuckDB is built for speed for analytics, so you can either have a single process that reads and writes, OR multiple read only processes. Reading and writing from multiple processes at the same time is not supported. This allows DuckDB to take advantage of RAM for much higher speeds.

We did just open source a way to read SQLite files as if they were a table in DuckDB, so if you can mix and match SQLite and DuckDB! Google sqlitescanner!

As for how things work in an R program, if your R program fails before you execute a commit (if you began with Begin transaction), then the next time you connect it should roll back your changes I believe.

Does that help a little? Thanks for trying DuckDB!

**Reply by Kodiologist** [reply]
Yes, that's helpful, thank you.

If multiple concurrent write attempts could lead to corruption or inconsistency, then I guess DuckDB isn't appropriate for e.g. a web application.

**Reply by Alex-Monahan** [reply]
Multiple concurrent write attempts from the same process are not an issue (multiple threads are fine). So if it's in the same process it can work, however a web app backend is not DuckDB's specialty or focus (as you guessed!).
